nomos-services/data-availability/sampling/src/network/adapters/libp2p.rs

```diff
@@ -0,0 +1,127 @@
+// std
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::pin::Pin;
+
+use rand::prelude::*;
+use rand_chacha::ChaCha20Rng;
+
+// crates
+use futures::{Stream, StreamExt};
+use libp2p_identity::PeerId;
+use tokio::sync::oneshot;
+// internal
+use crate::network::NetworkAdapter;
+use nomos_core::da::BlobId;
+use nomos_da_network_core::SubnetworkId;
+use nomos_da_network_service::backends::libp2p::validator::{
+    DaNetworkEvent, DaNetworkEventKind, DaNetworkMessage, DaNetworkValidatorBackend, SamplingEvent,
+};
+use nomos_da_network_service::{DaNetworkMsg, NetworkService};
+use overwatch_rs::services::relay::OutboundRelay;
+use overwatch_rs::services::ServiceData;
+use overwatch_rs::DynError;
+use subnetworks_assignations::MembershipHandler;
+
+#[derive(Debug, Clone)]
+pub struct DaNetworkSamplingSettings {
+    pub num_samples: u16,
+    pub subnet_size: SubnetworkId,
+}
+
+pub struct SamplingContext {
+    blob_id: BlobId,
+    subnets: Vec<SubnetworkId>,
+}
+
+pub struct Libp2pAdapter<Membership>
+where
+    Membership: MembershipHandler<NetworkId = SubnetworkId, Id = PeerId>
+        + Debug
+        + Clone
+        + Send
+        + Sync
+        + 'static,
+{
+    settings: DaNetworkSamplingSettings,
+    network_relay: OutboundRelay<
+        <NetworkService<DaNetworkValidatorBackend<Membership>> as ServiceData>::Message,
+    >,
+    pending_sampling: HashMap<BlobId, SamplingContext>,
+    // TODO: is there a better place for this? Do we need to have this even globally?
+    // Do we already have some source of randomness already?
+    rng: ChaCha20Rng,
```

Comment:
This is actually really important. And we should link it to the corresponding part in the spec, and if it is missing we should add it. (We can open an issue to track it).

Reply:
You are referring to the specification for randomness as such right? Not actually the content of the comment (where in the code it has to live).

In that case, yes, we should open an issue.

Reply:
Where it lives it doesn't matter much. Actually I think it could be spawn per request unless we need to use an specific seed from a source (I think it should not be the case). But questions like:
* What random algorithm to use?
* Do we need a common seed?
Need to be answered in the specification and matched in code (unless we consider is an implementation detail, which I'm not sure).